Documentation/design-docs/object-stack-allocation.md

```diff
@@ -0,0 +1,190 @@
+* Object field accesses may become cheaper if the compiler is able to do scalar replacement of the fields of the stack-allocated object
+(i.e., if the fields can be promoted).
+* Some field zero-initializations may be elided by the compiler.
+
+Object stack allocation is implemented in
+various Java runtimes.  This optimization is more important for Java since it doesn't have value types.
+
+## GitHub issues
+
+[roslyn #2104](https://github.com/dotnet/roslyn/issues/2104) Complier should optimize "alloc temporary small object" to "alloc on stack"
+
+[coreclr #1784](https://github.com/dotnet/coreclr/issues/1784) CLR/JIT should optimize "alloc temporary small object" to "alloc on stack" automatically
+
+## Escape Analysis
+
+An object is said to escape a method if it can be accessed after the method's execution has finished.
+An object allocation can be moved to the stack safely only if the object doesn't escape the allocating method.
+
+Several escape algorithms have been implemented in different Java implementations. Of the 3 algorithms listed in [references](References),
+[[1]](http://citeseerx.ist.psu.edu/viewdoc/download?doi=10.1.1.73.4799&rep=rep1&type=pdf)
+is the most precise and most expensive (it is based on connection graphs) and was used in the context of a static Java compiler,
+[[3]](https://pdfs.semanticscholar.org/1b33/dff471644f309392049c2791bca9a7f3b19c.pdf)
+is the least precise and cheapest (it doesn't track references through assignments of fields) and was used in MSR's Marmot implementation
+[[2]](https://www.usenix.org/legacy/events/vee05/full_papers/p111-kotzmann.pdf)
+is between
+[[1]](http://citeseerx.ist.psu.edu/viewdoc/download?doi=10.1.1.73.4799&rep=rep1&type=pdf) and 
+[[3]](https://pdfs.semanticscholar.org/1b33/dff471644f309392049c2791bca9a7f3b19c.pdf)
+both in analysis precision and cost. It was used in Java HotSpot.
+
+Effectiveness of object stack allocation depends in large part on whether escape analysis is done inter-procedurally.
+With intra-procedural analysis only, the compiler has to assume that arguments escape at all non-inlined call sites,
+which blocks many stack allocations. In particular, assuming that 'this' argument always escapes hurts the optimization.
+
+There are several choices for where escape analysis can be performed:
+
+### Analysis in the jit
+**Pros:**
+* The jit can analyze callee's code (subject to some restrictions, e.g., when running under profiler)
+  since there are no versioning considerations at jit time.
+* The optimization will apply to any msil code regardless of the language compiler or msil post-processing tools.
+* The jit already has IR that's suitable for escape analysis.
+
+**Cons:**
+* The jit analyzes methods top-down, i.e., callers before callees (when inlining), which doesn't fit well with the stack allocation optimization.
+* Full interprocedural analysis is too expensive for the jit, even at high tiering levels.
```

Comment:
has there been any consideration on 'less than full' interprocedural analysis.  For example, profiling-guided procedural analysis.  I know devs who have worked on jits where the runtime execution metrics were used to drive localized interprocedural analysis with very good results.  i.e. instead of doing "full" analysis, instead keep track of which procedures are very 'hot', as well as small relations between the hot procedures.  i.e. function A calls B calls C millions of times.  then go and analyze those small cliques.

Reply:
I think it's too strong a statement to say that "Full interprocedural analysis is too expensive for the jit, even at high tiering levels". If one assumes that with a higher-tiered JIT we would have the ability to memoize method properties with (in)validation, background on-demand/full interprocedural analysis would be feasible. That said, it's not likely to be practical any time soon.

Reply:
I re-worded the statement.